source/joker_effects.c

```diff
@@ -5,180 +5,206 @@
 #include "list.h"
 #include <stdlib.h>
 
-static JokerEffect default_joker_effect(Joker *joker, Card *scored_card) {
+
+#define SCORE_ON_CARD_SCORED_ONLY(scored_card, scored_when, effect) if (scored_when != JOKER_CALLBACK_ON_CARD_SCORED || !scored_card) { return effect; }
+#define SCORE_ON_CARD_HELD_ONLY(scored_card, scored_when, effect)   if (scored_when != JOKER_CALLBACK_ON_CARD_HELD   || !scored_card) { return effect; }
+#define SCORE_INDEPENDANT_ONLY(scored_card, scored_when, effect)    if (scored_when != JOKER_CALLBACK_INDEPENDANT)                    { return effect; }
+
+
+static JokerEffect default_joker_effect(Joker *joker, Card *scored_card, int scored_when) {
     JokerEffect effect = {0};
-    if (scored_card == NULL) effect.mult = 4;
+
+    SCORE_INDEPENDANT_ONLY(scored_card, scored_when, effect)
+
+    effect.mult = 4;
     return effect;
 }
 
-static JokerEffect sinful_joker_effect(Card *scored_card, u8 sinful_suit) {
+static JokerEffect sinful_joker_effect(Card *scored_card, u8 sinful_suit, int scored_when) {
     JokerEffect effect = {0};
-    if (scored_card != NULL && scored_card->suit == sinful_suit)
+
+    SCORE_ON_CARD_SCORED_ONLY(scored_card, scored_when, effect)
+
+    if (scored_card->suit == sinful_suit) {
         effect.mult = 3;
+    }
+
     return effect;
 }
 
-static JokerEffect greedy_joker_effect(Joker *joker, Card *scored_card) {
-    return sinful_joker_effect(scored_card, DIAMONDS);
+static JokerEffect greedy_joker_effect(Joker *joker, Card *scored_card, int scored_when) {
+    return sinful_joker_effect(scored_card, DIAMONDS, scored_when);
 }
 
-static JokerEffect lusty_joker_effect(Joker *joker, Card *scored_card) {
-    return sinful_joker_effect(scored_card, HEARTS);
+static JokerEffect lusty_joker_effect(Joker *joker, Card *scored_card, int scored_when) {
+    return sinful_joker_effect(scored_card, HEARTS, scored_when);
 }
 
-static JokerEffect wrathful_joker_effect(Joker *joker, Card *scored_card) {
-    return sinful_joker_effect(scored_card, SPADES);
+static JokerEffect wrathful_joker_effect(Joker *joker, Card *scored_card, int scored_when) {
+    return sinful_joker_effect(scored_card, SPADES, scored_when);
 }
 
-static JokerEffect gluttonous_joker_effect(Joker *joker, Card *scored_card) {
-    return sinful_joker_effect(scored_card, CLUBS);
+static JokerEffect gluttonous_joker_effect(Joker *joker, Card *scored_card, int scored_when) {
+    return sinful_joker_effect(scored_card, CLUBS, scored_when);
 }
 
-static JokerEffect jolly_joker_effect(Joker *joker, Card *scored_card) {
+static JokerEffect jolly_joker_effect(Joker *joker, Card *scored_card, int scored_when) {
     JokerEffect effect = {0};
-    if (scored_card != NULL)
-        return effect; // if card != null, we are not at the end-phase of scoring yet
 
+    SCORE_INDEPENDANT_ONLY(scored_card, scored_when, effect)
```

Comment:
```suggestion
    SCORE_ON_EVENT_ONLY(JOKER_CALLBACK_INDEPENDANT, scored_when, effect)
```

Batch - Consolidate joker event restriction macros